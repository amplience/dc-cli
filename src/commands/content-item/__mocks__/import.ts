import { ImportItemBuilderOptions } from '../../../interfaces/import-item-builder-options.interface';
import { ConfigurationParameters } from '../../configure';
import { Arguments } from 'yargs';

type ReturnType = boolean | 'throw';
let expectedReturn: ReturnType = true;

export const calls: Arguments<ImportItemBuilderOptions & ConfigurationParameters>[] = [];
export const setExpectedReturn = (value: ReturnType): void => {
  expectedReturn = value;
};

export const handler = async (
  argv: Arguments<ImportItemBuilderOptions & ConfigurationParameters>
): Promise<boolean> => {
  calls.push(argv);

  if (expectedReturn == 'throw') {
    throw new Error('Forced throw in test.');
  }

  return expectedReturn;
};
