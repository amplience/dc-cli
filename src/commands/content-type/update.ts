import DataPresenter, { RenderingOptions, RenderingArguments } from '../../view/data-presenter';

import { ConfigurationParameters } from '../configure';
import { Arguments, Argv } from 'yargs';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { ContentType, ContentTypeVisualization, ContentTypeIcon, ContentTypeCard } from 'dc-management-sdk-js';
import { transformYargObjectToArray, YargObject } from '../../common/yargs/yargs-object-transformer';
import { singleItemTableOptions } from '../../common/table/table.consts';

export const command = 'update [id]';

export const desc = 'Update a Content Type';

export const builder = (yargs: Argv): void => {
  yargs
    .positional('id', {
      type: 'string',
      demandOption: true,
      describe: 'content-type ID'
    })
    .options({
      label: {
        type: 'string',
        describe: 'content-type label'
      },
      icons: {
        describe: 'content-type icons'
      },
      visualizations: {
        describe: 'content-type visualizations'
      },
      cards: {
        describe: 'content-type cards'
      },
      ...RenderingOptions
    });
};

interface ContentTypeUpdateBuilderOptions {
  id: string;
  label?: string;
  icons?: YargObject<ContentTypeIcon> | boolean;
  cards?: YargObject<ContentTypeCard> | boolean;
  visualizations?: YargObject<ContentTypeVisualization> | boolean;
}

export const handler = async (
  argv: Arguments<ContentTypeUpdateBuilderOptions & ConfigurationParameters & RenderingArguments>
): Promise<void> => {
  const client = dynamicContentClientFactory(argv);
  const { id, label, icons, visualizations, cards } = argv;
  const contentType = await client.contentTypes.get(id);
  const mutatedContentType = new ContentType({
    settings: {
      ...contentType.settings,
      ...(label ? { label } : {}),
      ...(icons ? { icons: transformYargObjectToArray<ContentTypeIcon>(icons) } : {}),
      ...(cards ? { cards: transformYargObjectToArray<ContentTypeCard>(cards) } : {}),
      ...(visualizations
        ? {
            visualizations: transformYargObjectToArray<ContentTypeVisualization>(visualizations)
          }
        : {})
    },
    _links: contentType._links
  });
  const updatedContentType = await contentType.related.update(mutatedContentType);

  new DataPresenter(updatedContentType.toJson()).render({
    json: argv.json,
    tableUserConfig: singleItemTableOptions
  });
};
