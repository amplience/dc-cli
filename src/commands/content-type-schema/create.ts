import { CommandOptions } from '../../interfaces/command-options.interface';
import DataPresenter, { RenderingArguments, RenderingOptions } from '../../view/data-presenter';
import { Arguments } from 'yargs';
import { ConfigurationParameters } from '../configure';
import { ValidationLevel } from 'dc-management-sdk-js';
import { singleItemTableOptions } from '../../common/table/table.consts';
import { createContentTypeSchema } from './create.service';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { jsonResolver } from '../../common/import/json-resolver';

export const command = 'create';

export const desc = 'Create Content Type Schema';

export const builder: CommandOptions = {
  schema: {
    type: 'string',
    demandOption: true,
    description: 'content-type-schema Source Location',
    requiresArg: true
  },
  validationLevel: {
    type: 'string',
    choices: Object.values(ValidationLevel),
    demandOption: true,
    description: 'content-type-schema Validation Level',
    requiresArg: true
  },
  ...RenderingOptions
};

export interface BuilderOptions {
  schema: string;
  validationLevel: string;
}

export const handler = async (
  argv: Arguments<BuilderOptions & ConfigurationParameters & RenderingArguments>
): Promise<void> => {
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const schemaBody = await jsonResolver(argv.schema);
  const contentTypeSchemaResult = await createContentTypeSchema(
    schemaBody,
    argv.validationLevel as ValidationLevel,
    hub
  );

  return new DataPresenter(contentTypeSchemaResult.toJSON()).render({
    json: argv.json,
    tableUserConfig: singleItemTableOptions
  });
};
