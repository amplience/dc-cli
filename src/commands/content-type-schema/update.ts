import DataPresenter, { RenderingArguments, RenderingOptions } from '../../view/data-presenter';
import { Arguments, Argv } from 'yargs';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { ConfigurationParameters } from '../configure';
import { ValidationLevel } from 'dc-management-sdk-js';
import { jsonResolver } from '../../common/import/json-resolver';
import { singleItemTableOptions } from '../../common/table/table.consts';
import { updateContentTypeSchema } from './update.service';

export const command = 'update <id>';

export const desc = 'Update Content Type Schema';

export const builder = (yargs: Argv): void => {
  yargs
    .positional('id', {
      describe: 'Content Type Schema ID',
      type: 'string'
    })
    .options({
      schema: {
        type: 'string',
        demandOption: true,
        description: 'Content Type Schema Source Location',
        requiresArg: true
      },
      validationLevel: {
        type: 'string',
        choices: Object.values(ValidationLevel),
        demandOption: true,
        description: 'Content Type Schema Validation Level',
        requiresArg: true
      },
      ...RenderingOptions
    });
};

export interface BuilderOptions {
  id: string;
  schema: string;
  validationLevel: ValidationLevel;
}

export const handler = async (
  argv: Arguments<BuilderOptions & ConfigurationParameters & RenderingArguments>
): Promise<void> => {
  const { id, schema, validationLevel } = argv;
  const client = dynamicContentClientFactory(argv);
  const schemaBody = await jsonResolver(schema);
  const contentTypeSchema = await client.contentTypeSchemas.get(id);
  const contentTypeSchemaResult = await updateContentTypeSchema(contentTypeSchema, schemaBody, validationLevel);

  new DataPresenter(contentTypeSchemaResult.toJSON()).render({
    json: argv.json,
    tableUserConfig: singleItemTableOptions
  });
};
