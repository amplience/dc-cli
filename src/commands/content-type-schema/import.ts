import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import { ContentTypeSchema, DynamicContent, Hub, ValidationLevel } from 'dc-management-sdk-js';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { createStream } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { TableStream } from '../../interfaces/table.interface';
import { ImportBuilderOptions } from '../../interfaces/import-builder-options.interface';
import chalk from 'chalk';
import { createContentTypeSchema } from './create.service';
import { updateContentTypeSchema } from './update.service';
import { ImportResult, loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import { jsonResolver } from '../../common/import/json-resolver';

export const command = 'import <dir>';

export const desc = 'Import Content Type Schemas';

export interface SchemaOptions {
  validation: ValidationLevel;
}

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    describe: 'Directory containing Content Type Schema definitions',
    type: 'string'
  });
};

export const storedSchemaMapper = (
  schema: ContentTypeSchema,
  storedSchemas: ContentTypeSchema[]
): ContentTypeSchema => {
  const found = storedSchemas.find(stored => stored.schemaId === schema.schemaId);
  const mutatedSchema = found ? { ...schema, id: found.id } : schema;

  return new ContentTypeSchema(mutatedSchema);
};

export const doCreate = async (hub: Hub, schema: ContentTypeSchema): Promise<ContentTypeSchema> => {
  try {
    const createdSchemaType = await createContentTypeSchema(
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE,
      hub
    );

    return createdSchemaType;
  } catch (err) {
    throw new Error(`Error registering content type schema with body: ${schema.body}\n\n${err}`);
  }
};

const equals = (a: ContentTypeSchema, b: ContentTypeSchema): boolean =>
  a.id === b.id && a.schemaId === b.schemaId && a.body === b.body;

export const doUpdate = async (
  client: DynamicContent,
  schema: ContentTypeSchema
): Promise<{ contentTypeSchema: ContentTypeSchema; updateStatus: UpdateStatus }> => {
  try {
    const retrievedSchema = await client.contentTypeSchemas.get(schema.id || '');
    if (equals(retrievedSchema, schema)) {
      return { contentTypeSchema: retrievedSchema, updateStatus: UpdateStatus.SKIPPED };
    }
    const updatedSchema = await updateContentTypeSchema(
      retrievedSchema,
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE
    );

    return { contentTypeSchema: updatedSchema, updateStatus: UpdateStatus.UPDATED };
  } catch (err) {
    throw new Error(`Error updating content type schema ${schema.schemaId || '<unknown>'}: ${err.message}`);
  }
};

export const processSchemas = async (
  schemasToProcess: ContentTypeSchema[],
  client: DynamicContent,
  hub: Hub
): Promise<void> => {
  const tableStream = (createStream(streamTableOptions) as unknown) as TableStream;

  tableStream.write([chalk.bold('ID'), chalk.bold('Schema ID'), chalk.bold('Result')]);
  for (const schema of schemasToProcess) {
    let schemaId = schema.id;
    let status: ImportResult;
    if (schema.id) {
      const result = await doUpdate(client, schema);
      status = result.updateStatus === UpdateStatus.SKIPPED ? 'UP-TO-DATE' : 'UPDATED';
    } else {
      const result = await doCreate(hub, schema);
      schemaId = result.id || 'UNKNOWN';
      status = 'CREATED';
    }
    tableStream.write([schemaId || '', schema.schemaId || '', status]);
  }
  process.stdout.write('\n');
};

export const resolveSchemaBody = async (schemas: ContentTypeSchema[], dir: string): Promise<ContentTypeSchema[]> => {
  for (const schema of schemas) {
    if (schema.body) {
      schema.body = await jsonResolver(schema.body, dir);
    }
  }
  return schemas;
};

export const handler = async (argv: Arguments<ImportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir } = argv;
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const schemas = Object.values(loadJsonFromDirectory<ContentTypeSchema>(dir, ContentTypeSchema));
  await resolveSchemaBody(schemas, dir);
  const storedSchemas = await paginator(hub.related.contentTypeSchema.list);
  const schemasToProcess = schemas.map(schemas => storedSchemaMapper(schemas, storedSchemas));

  await processSchemas(schemasToProcess, client, hub);
};
