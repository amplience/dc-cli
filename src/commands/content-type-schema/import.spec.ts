import Yargs = require('yargs/yargs');

import * as importModule from './import';
import {
  command,
  builder,
  handler,
  storedSchemaMapper,
  processSchemas,
  doCreate,
  doUpdate,
  LOG_FILENAME
} from './import';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { ContentTypeSchema, ValidationLevel, Hub } from 'dc-management-sdk-js';
import { table } from 'table';
import { createContentTypeSchema } from './create.service';
import { updateContentTypeSchema } from './update.service';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import { resolveSchemaBody } from '../../services/resolve-schema-body';
import { FileLog } from '../../common/file-log';
import { streamTableOptions } from '../../common/table/table.consts';
import chalk from 'chalk';

jest.mock('fs');
jest.mock('table');
jest.mock('../../common/dc-management-sdk-js/paginator');
jest.mock('../../services/dynamic-content-client-factory');
jest.mock('../../services/import.service');
jest.mock('../../services/resolve-schema-body');
jest.mock('./create.service');
jest.mock('./update.service');

const schemaId = 'https://schema.localhost.com/test-1.json';
const schemaBodyJson =
  '{\n\t"$schema": "http://json-schema.org/draft-07/schema#",\n\t"$id": "https://schema.localhost.com/test-1.json",\n\n\t"title": "Test Schema 1",\n\t"description": "Test Schema 1",\n\n\t"allOf": [\n\t\t{\n\t\t\t"$ref": "http://bigcontent.io/cms/schema/v1/core#/definitions/content"\n\t\t}\n\t],\n\t\n\t"type": "object",\n\t"properties": {\n\t\t\n\t},\n\t"propertyOrder": []\n}';

describe('content-type-schema import command', (): void => {
  afterEach((): void => {
    jest.resetAllMocks();
  });

  it('should implement an import command', () => {
    expect(command).toEqual('import <dir>');
  });

  describe('builder tests', () => {
    it('should configure yargs', () => {
      const argv = Yargs(process.argv.slice(2));
      const spyPositional = jest.spyOn(argv, 'positional').mockReturnThis();
      const spyOption = jest.spyOn(argv, 'option').mockReturnThis();
      builder(argv);

      expect(spyPositional).toHaveBeenCalledWith('dir', {
        describe: 'Directory containing Content Type Schema definitions',
        type: 'string'
      });

      expect(spyOption).toHaveBeenCalledWith('logFile', {
        type: 'string',
        default: LOG_FILENAME,
        describe: 'Path to a log file to write to.'
      });
    });
  });

  describe('storedSchemaMapper', () => {
    it('should map a stored schema to the imported list with matching imported schema', () => {
      const schemaBody = { $id: 'schema-id' };
      const importedSchema = new ContentTypeSchema({
        body: JSON.stringify(schemaBody),
        validationLevel: ValidationLevel.CONTENT_TYPE,
        schemaId: schemaBody.$id
      });
      const storedContentTypeSchema = new ContentTypeSchema({
        id: 'stored-id',
        schemaId: schemaBody.$id,
        ...importedSchema.toJSON()
      });
      const storedSchemaList = [storedContentTypeSchema];
      const result = storedSchemaMapper(importedSchema, storedSchemaList);

      expect(result).toEqual(expect.objectContaining(storedContentTypeSchema.toJSON()));
      expect(result.id).toBeDefined();
    });

    it('should return the imported schema when there is no matching stored schema', () => {
      const schemaBody = { $id: 'schema-id' };
      const importedSchema = new ContentTypeSchema({
        body: JSON.stringify(schemaBody),
        validationLevel: ValidationLevel.CONTENT_TYPE,
        schemaId: schemaBody.$id
      });
      const storedContentTypeSchema = new ContentTypeSchema({
        schemaId: 'stored-schema-id',
        ...importedSchema.toJSON()
      });
      const storedSchemaList = [storedContentTypeSchema];
      const result = storedSchemaMapper(importedSchema, storedSchemaList);

      expect(result).toEqual(expect.objectContaining(importedSchema.toJSON()));
      expect(result.id).toBeUndefined();
    });
  });

  describe('doCreate', () => {
    it('should create a content type schema and report the results', async () => {
      const hub = new Hub();
      const log = new FileLog();

      const contentTypeSchema = {
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE
      } as ContentTypeSchema;
      (createContentTypeSchema as jest.Mock).mockResolvedValueOnce({ ...contentTypeSchema, id: 'create-id', schemaId });
      const result = await doCreate(hub, contentTypeSchema, log);

      expect(createContentTypeSchema).toHaveBeenCalledWith(
        contentTypeSchema.body,
        contentTypeSchema.validationLevel,
        hub
      );
      expect(result).toEqual({ ...contentTypeSchema, id: 'create-id', schemaId });
      expect(log.getData('CREATE')).toMatchInlineSnapshot(`
        Array [
          "create-id",
        ]
      `);
    });

    it('should throw an error when content type schema fails to create', async () => {
      const hub = new Hub();
      const log = new FileLog();

      const contentTypeSchema = {
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE
      } as ContentTypeSchema;
      (createContentTypeSchema as jest.Mock).mockImplementation(() => {
        throw new Error('Error creating content type schema');
      });

      await expect(doCreate(hub, contentTypeSchema, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('CREATE')).toEqual([]);
    });
  });

  describe('doUpdate', () => {
    const mockGetContentTypeSchema = jest.fn();

    beforeEach(() => {
      (dynamicContentClientFactory as jest.Mock).mockReturnValue({
        contentTypeSchemas: {
          get: mockGetContentTypeSchema
        }
      });
    });
    it('should update a content type schema and report the results', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const log = new FileLog();

      const storedContentTypeSchema = {
        id: 'stored-id',
        schemaId,
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE,
        version: 1
      } as ContentTypeSchema;
      const mutatedContentTypeSchema = {
        ...storedContentTypeSchema,
        body: `{\n\t"$schema": "http://json-schema.org/draft-07/schema#",\n\t"$id": "${schemaId}",\n\n\t"title": "Test Schema 1 - updated",\n\t"description": "Test Schema 1- updated",\n\n\t"allOf": [\n\t\t{\n\t\t\t"$ref": "http://bigcontent.io/cms/schema/v1/core#/definitions/content"\n\t\t}\n\t],\n\t\n\t"type": "object",\n\t"properties": {\n\t\t\n\t},\n\t"propertyOrder": []\n}`,
        version: 2
      } as ContentTypeSchema;
      mockGetContentTypeSchema.mockResolvedValueOnce(new ContentTypeSchema(storedContentTypeSchema));
      (updateContentTypeSchema as jest.Mock).mockResolvedValueOnce({
        ...mutatedContentTypeSchema,
        id: 'stored-id',
        schemaId
      });
      const result = await doUpdate(client, mutatedContentTypeSchema, log);

      expect(log.getData('UPDATE')).toMatchInlineSnapshot(`
        Array [
          "stored-id 1 2",
        ]
      `);

      expect(updateContentTypeSchema).toHaveBeenCalledWith(
        expect.objectContaining(storedContentTypeSchema),
        mutatedContentTypeSchema.body,
        mutatedContentTypeSchema.validationLevel
      );
      expect(result).toEqual({ contentTypeSchema: mutatedContentTypeSchema, updateStatus: UpdateStatus.UPDATED });
    });

    it('should update a content type when only the validationLevel has been updated', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const log = new FileLog();
      const storedContentTypeSchema = {
        id: 'stored-id',
        schemaId,
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE,
        version: 1
      } as ContentTypeSchema;
      const mutatedContentTypeSchema = {
        ...storedContentTypeSchema,
        validationLevel: ValidationLevel.SLOT,
        version: 2
      } as ContentTypeSchema;
      mockGetContentTypeSchema.mockResolvedValueOnce(new ContentTypeSchema(storedContentTypeSchema));
      (updateContentTypeSchema as jest.Mock).mockResolvedValueOnce({
        ...mutatedContentTypeSchema,
        id: 'stored-id',
        schemaId
      });
      const result = await doUpdate(client, mutatedContentTypeSchema, log);

      expect(log.getData('UPDATE')).toMatchInlineSnapshot(`
        Array [
          "stored-id 1 2",
        ]
      `);
      expect(updateContentTypeSchema).toHaveBeenCalledWith(
        expect.objectContaining(storedContentTypeSchema),
        mutatedContentTypeSchema.body,
        mutatedContentTypeSchema.validationLevel
      );
      expect(result).toEqual({ contentTypeSchema: mutatedContentTypeSchema, updateStatus: UpdateStatus.UPDATED });
    });

    it('should skip updating a content type schema when no changes detected and report the results', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const log = new FileLog();

      const storedContentTypeSchema = {
        id: 'stored-id',
        schemaId,
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE
      } as ContentTypeSchema;
      const mutatedContentTypeSchema = {
        ...storedContentTypeSchema,
        body: schemaBodyJson
      } as ContentTypeSchema;
      mockGetContentTypeSchema.mockResolvedValueOnce(new ContentTypeSchema(storedContentTypeSchema));

      const result = await doUpdate(client, mutatedContentTypeSchema, log);

      expect(log.getData('UPDATE')).toEqual([]);
      expect(updateContentTypeSchema).toHaveBeenCalledTimes(0);
      expect(result).toEqual(expect.objectContaining({ updateStatus: UpdateStatus.SKIPPED }));
    });

    it('should throw an error when content type schema fails to create', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const log = new FileLog();

      const contentTypeSchema = {
        id: 'stored-id',
        schemaId,
        body: schemaBodyJson,
        validationLevel: ValidationLevel.CONTENT_TYPE
      } as ContentTypeSchema;
      mockGetContentTypeSchema.mockImplementationOnce(() => {
        throw new Error('Error getting content type schema');
      });
      await expect(doUpdate(client, contentTypeSchema, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
    });
  });

  describe('processSchemas', () => {
    let mockTable: jest.Mock;

    beforeEach(() => {
      mockTable = table as jest.Mock;
      mockTable.mockImplementation(jest.requireActual('table').table);
    });

    it('should successfully create and update a schema', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const hub = new Hub();

      const contentTypeSchemaToCreate = {
        schemaId,
        body: `{\n\t"$schema": "http://json-schema.org/draft-07/schema#",\n\t"$id": "${schemaId}",\n\n\t"title": "Test Schema - Updated",\n\t"description": "Test Schema - Updated",\n\n\t"allOf": [\n\t\t{\n\t\t\t"$ref": "http://bigcontent.io/cms/schema/v1/core#/definitions/content"\n\t\t}\n\t],\n\t\n\t"type": "object",\n\t"properties": {\n\t\t\n\t},\n\t"propertyOrder": []\n}`,
        validationLevel: ValidationLevel.CONTENT_TYPE
      } as ContentTypeSchema;
      const contentTypeSchemaToUpdate = { ...contentTypeSchemaToCreate, id: 'stored-id' } as ContentTypeSchema;
      const schemasToProcess = [contentTypeSchemaToCreate, contentTypeSchemaToUpdate];

      jest
        .spyOn(importModule, 'doCreate')
        .mockResolvedValueOnce({ ...contentTypeSchemaToCreate, id: 'new-id' } as ContentTypeSchema);
      jest
        .spyOn(importModule, 'doUpdate')
        .mockResolvedValueOnce({ contentTypeSchema: contentTypeSchemaToUpdate, updateStatus: UpdateStatus.UPDATED });

      await processSchemas(schemasToProcess, client, hub, new FileLog());

      expect(importModule.doCreate).toHaveBeenCalledWith(hub, contentTypeSchemaToCreate, expect.any(FileLog));
      expect(importModule.doUpdate).toHaveBeenCalledWith(client, contentTypeSchemaToUpdate, expect.any(FileLog));
      expect(mockTable).toHaveBeenCalledTimes(1);
      expect(mockTable).toHaveBeenNthCalledWith(
        1,
        [
          [chalk.bold('ID'), chalk.bold('Schema ID'), chalk.bold('Result')],
          ['new-id', schemaId, 'CREATED'],
          ['stored-id', schemaId, 'UPDATED']
        ],
        streamTableOptions
      );
    });
  });

  describe('handler tests', () => {
    const yargArgs = {
      $0: 'test',
      _: ['test']
    };
    const config = {
      clientId: 'client-id',
      clientSecret: 'client-id',
      hubId: 'hub-id'
    };
    const mockGetHub = jest.fn();
    const argv = {
      ...yargArgs,
      ...config,
      dir: 'my-dir'
    };

    beforeEach(() => {
      (dynamicContentClientFactory as jest.Mock).mockReturnValue({
        hubs: {
          get: mockGetHub
        }
      });
    });

    it('should process schemas from a local directory', async (): Promise<void> => {
      const schemaToCreate = {
        body: `{\n\t"$schema": "http://json-schema.org/draft-07/schema#",\n\t"$id": "https://schema.localhost.com/remote-test-1.json",\n\n\t"title": "Test Schema 1",\n\t"description": "Test Schema 1",\n\n\t"allOf": [\n\t\t{\n\t\t\t"$ref": "http://bigcontent.io/cms/schema/v1/core#/definitions/content"\n\t\t}\n\t],\n\t\n\t"type": "object",\n\t"properties": {\n\t\t\n\t},\n\t"propertyOrder": []\n}`,
        validationLevel: ValidationLevel.CONTENT_TYPE,
        schemaId: 'create-schema-id'
      };
      const schemaToUpdate = {
        body: `{\n\t"$schema": "http://json-schema.org/draft-07/schema#",\n\t"$id": "https://schema.localhost.com/local-test-2.json",\n\n\t"title": "Test Schema 2",\n\t"description": "Test Schema 2",\n\n\t"allOf": [\n\t\t{\n\t\t\t"$ref": "http://bigcontent.io/cms/schema/v1/core#/definitions/content"\n\t\t}\n\t],\n\t\n\t"type": "object",\n\t"properties": {\n\t\t\n\t},\n\t"propertyOrder": []\n}`,
        validationLevel: ValidationLevel.CONTENT_TYPE,
        schemaId: 'update-schema-id'
      };

      mockGetHub.mockResolvedValue(new Hub());
      const loadJsonFromDirectoryResult = { 'dir/create.json': schemaToCreate, 'dir/update.json': schemaToUpdate };
      (loadJsonFromDirectory as jest.Mock).mockReturnValueOnce(loadJsonFromDirectoryResult);
      (resolveSchemaBody as jest.Mock).mockImplementation(args => [args, {}]);
      (paginator as jest.Mock).mockResolvedValue([{ ...schemaToUpdate, id: 'stored-id' }]);

      const processSchemasSpy = jest
        .spyOn(importModule, 'processSchemas')
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        .mockImplementation(async (): Promise<void> => {});

      await handler(argv);

      expect(resolveSchemaBody as jest.Mock).toHaveBeenCalledWith(loadJsonFromDirectoryResult, argv.dir);
      expect(loadJsonFromDirectory).toHaveBeenCalledWith('my-dir', ContentTypeSchema);
      expect(paginator).toHaveBeenCalledWith(expect.any(Function));
      expect(processSchemasSpy).toHaveBeenCalledWith(
        [expect.objectContaining(schemaToCreate), expect.objectContaining({ ...schemaToUpdate, id: 'stored-id' })],
        expect.any(Object),
        expect.any(Object),
        expect.any(Object)
      );
    });

    it('should report all resolve schema body errors', async () => {
      mockGetHub.mockResolvedValue(new Hub());
      (loadJsonFromDirectory as jest.Mock).mockReturnValueOnce({});
      (resolveSchemaBody as jest.Mock).mockResolvedValue([
        {},
        { 'filename.json': new Error('Unable to resolve filename.json') }
      ]);
      await expect(handler(argv)).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
