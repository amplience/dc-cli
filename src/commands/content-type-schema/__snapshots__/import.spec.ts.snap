// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content-type-schema import command doCreate should throw an error when content type schema fails to create 1`] = `
"Error registering content type schema with body: {
	\\"$schema\\": \\"http://json-schema.org/draft-04/schema#\\",
	\\"id\\": \\"https://schema.localhost.com/test-1.json\\",

	\\"title\\": \\"Test Schema 1\\",
	\\"description\\": \\"Test Schema 1\\",

	\\"allOf\\": [
		{
			\\"$ref\\": \\"http://bigcontent.io/cms/schema/v1/core#/definitions/content\\"
		}
	],
	
	\\"type\\": \\"object\\",
	\\"properties\\": {
		
	},
	\\"propertyOrder\\": []
}

Error: Error creating content type schema"
`;

exports[`content-type-schema import command doUpdate should throw an error when content type schema fails to create 1`] = `"Error updating content type schema https://schema.localhost.com/test-1.json: Error getting content type schema"`;
